src/tree-sitter-language-mode.js

```diff
@@ -947,55 +950,97 @@ class HighlightIterator {
       }
     );
 
-    this.iterators = [
-      this.languageMode.rootLanguageLayer.buildHighlightIterator()
-    ];
-    for (const marker of injectionMarkers) {
-      this.iterators.push(marker.languageLayer.buildHighlightIterator());
-    }
-    this.iterators.sort((a, b) => b.getIndex() - a.getIndex());
-
     const containingTags = [];
     const containingTagStartIndices = [];
     const targetIndex = this.languageMode.buffer.characterIndexForPosition(
       targetPosition
     );
-    for (let i = this.iterators.length - 1; i >= 0; i--) {
-      this.iterators[i].seek(
-        targetIndex,
-        containingTags,
-        containingTagStartIndices
-      );
+
+    this.iterators = [];
+    const iterator = this.languageMode.rootLanguageLayer.buildHighlightIterator();
+    if (iterator.seek(targetIndex, containingTags, containingTagStartIndices)) {
+      this.iterators.push(iterator);
     }
-    this.iterators.sort((a, b) => b.getIndex() - a.getIndex());
+    for (const marker of injectionMarkers) {
+      const iterator = marker.languageLayer.buildHighlightIterator();
+      if (
+        iterator.seek(targetIndex, containingTags, containingTagStartIndices)
+      ) {
+        this.iterators.push(iterator);
+      }
+    }
+    this.iterators.sort((a, b) => b.compare(a));
     return containingTags;
   }
 
   moveToSuccessor() {
-    const lastIndex = this.iterators.length - 1;
-    const leader = this.iterators[lastIndex];
-    leader.moveToSuccessor();
-    const leaderCharIndex = leader.getIndex();
-    let i = lastIndex;
-    while (i > 0 && this.iterators[i - 1].getIndex() < leaderCharIndex) i--;
-    if (i < lastIndex) this.iterators.splice(i, 0, this.iterators.pop());
+    let leader = last(this.iterators);
+    if (leader.moveToSuccessor()) {
+      const leaderIndex = this.iterators.length - 1;
+      let i = leaderIndex;
+      while (i > 0 && this.iterators[i - 1].compare(leader) < 0) i--;
+      if (i < leaderIndex) {
+        this.iterators.splice(i, 0, this.iterators.pop());
+        this.trackLayerDepth();
+      }
+    } else {
+      this.iterators.pop();
+      this.trackLayerDepth();
+    }
+  }
```

Comment:
The main difference in `moveToSuccessor` is that like `seek`, I've updated it so that it can return `false`, meaning that the iterator has fallen off the end and should be removed.

Reply:
Yeah, it's ok for `moveToSuccessor` to not actually change position, but to just update the opening and closing scopes. In the `TreeSitterLanguageMode`, this actually happens move frequently -  it happens whenever one scope starts at the same position where another scope ends. For simplicity of implementation, we don't report the closing and opening scopes in the same iteration.